Add `init(...) async throws` overloads to the `Store`.

rl-pavel opened the pull request:
This change adds 2 ways to observe/await for `items` to be loaded in the Store: 
- via the new `async` `Store.init` overload
- by using `await store.itemsHaveLoaded()` before accessing `store.items`

## Why
I ran into an issue after implementing a thin `Store` dependency wrapper to use with [The Composable Architecture](https://github.com/pointfreeco/swift-composable-architecture), where I tried loading `store.items` on launch, but it often returned an empty array. 

I believe part of the issue is that TCA seems to load `@Dependeny` lazily, so accessing the `items` right after init doesn't work. One workaround I found is to add `_ = someStore` to the root reducer's `init`. That worked better, however I worry that there might still be a race if/when the stored data grows and takes longer to load. 

## Compatibility

If the store was constructed in an `async` context, the compiler will try to use the new `async` version of the `init` and return errors about missing `try` and `await`. This could be addressed with different naming, but I'm not sure what could be used here. Thoughts? 🤔

mergesort commented:
Hey @rl-pavel, thanks for the contribution, I always appreciate when someone does the research and work. 🙂

I've actually tried to tackle the idea of making `init` async a few times, and haven't come up with a satisfactory option. It was discussed in [this thread](https://github.com/mergesort/Boutique/discussions/26#discussioncomment-3303736), the biggest issues being these two issues.

> 1. You can't use an async property in a property wrapper, so you can't use this Store in the @Stored property wrapper.
> 2. You can't use this in a global context, so static let wouldn't work either.

I'm absolutely open to solutions, but don't want to really break the really nice API this allows for using `@Stored`. If there's a way to work around those issues I would definitely consider this option.

Now in terms of a workaround, I don't have a great answer for you nor do I love this but to avoid showing an empty state while the async function kicks off I've been doing is storing the _latest state_ in a `@StoredValue` which is synchronous because `UserDefaults` gets to cheat and access data in a synchronous manner.

I will admit that it's been too long since I used TCA (especially with the new async/await APIs) that I can't answer the TCA-specific portion of your question, but hopefully that's enough information to further the conversation!

rl-pavel commented:
@mergesort thanks the context, I have missed that thread. 

Since this PR adds the init as an overload, i.e. the synchronous one with the background task still exists, I don't think this breaks the use case you're describing for `@Stored`? Unless I'm missing something else here.

As for my issue, I've updated the PR to store and expose the `Task` that loads the items so it can be `await`ed on. It is still run in the background in the syncronous `init`, and the `async` init version `await`s for the completion at the end. What do you think of this approach?

mergesort commented:
Hey @rl-pavel, I finally got a chance to take a closer look and I think it looks pretty good! I'll be upfront and say that I'll need to really try and use this to merge it in and have been a little underwater with work recently. In the mean time I'm going to  add a little bit more information below, and some individual comments on the pull request that I think would benefit from locality.

One thing I'd love to see added if you don't mind is high level documentation about when to use the sync or async versions, along with the limitations of each approach. I know it's not a trivial amount of work, but in the context of an open source project others depend on I think it's important to do before merging in the pull request with a big context change. The places I'm thinking it would be valuable to add this to are the readme, and the two docc articles, "The @Stored Family Of Property Wrappers" and "Using Stores".

I'd especially love to see this because even as the library author I'm now having a harder time reasoning about the different combinations and possibilities, when I should be using the sync or async version (especially in the context of dropping it into an actual app). Not knowing the pros, cons, and benefits of each I can only imagine the perspective of someone coming to, which is why I think enumerating them would be really helpful.

Thanks again for the work so far, it's really neat and I appreciate you tackling this problem that I've struggled to figure out a clean solution to before!

mergesort commented on Tests/BoutiqueTests/StoreTests.swift:
```diff
@@ -8,7 +8,7 @@ final class StoreTests: XCTestCase {
 
     override func setUp() async throws {
-        store = Store<BoutiqueItem>(
+        store = try await Store<BoutiqueItem>(
```
Would it be possible to make a separate `asyncStore` property for the tests that require an async Store? Ideally any tests where the behavior may differ significantly would have their own tests in case a change is made that introduces a bug only when the Store is synchronous, similar to how `StoredValue` and `AsyncStoredValue` have their own suite of tests.

rl-pavel replied:
Hm, this one's tricky. The functionality in both async/non-async init cases would be the same, the only difference is that the later returns before `items` are retrieved. However since `setUp` also does a `removeAll()` it's not really relevant here.

I've [written a test](https://github.com/mergesort/Boutique/pull/39/commits/607e7ccb68f6d74305c3eda15349b8e713894309) that checks that creating a new store with `await` will have the `items` loaded. That test would fail when using a non-async init. However it's also tricky to test given that the Swift compiler prefers using the `async` version of the overload in an `async` context. In any case, I'm not sure if testing that behavior would be useful, since trying to fetch `items` that have not loaded is probably not desirable 😅

mergesort replied:
What I was thinking was was to have two different Stores, one for async tests and one for sync tests. That way you wouldn't have to make assumptions about whether the store was sync or async, you just have two Stores and two sets of tests for a lot of the functionality.

I don't think it's necessary to rewrite _every test_, but I do think it would be good to have some baseline of tests that consider when the functionality could potentially differ, maybe such as one or two of the chained operation tests, `testInsertingItems()`, `testReadingItems()`, `testRemovingItems()`, and `testPublishedItemsSubscription()`.

mergesort commented on Sources/Boutique/Store.swift:
```diff
@@ -65,16 +76,20 @@ public final class Store<Item: Codable & Equatable>: ObservableObject {
+    ///   - storage: A `StorageEngine` to initialize a ``Store`` instance with.
+    ///   - cacheIdentifier: A `KeyPath` from the `Item` pointing to a `String`, which the ``Store``
+    ///   will use to create a unique identifier for the item when it's saved.
+    @MainActor public init(storage: StorageEngine, cacheIdentifier: KeyPath<Item, String>) async throws {
```
Nit/total personal preference: For functions I have a preference to putting annotations such as `@MainActor` on its own line above the function. For reference you can see that in the usage of `@_disfavoredOverload` in `Store.swift` as well.

mergesort commented on Sources/Boutique/Store.swift:
```diff
@@ -54,9 +54,20 @@ public final class Store<Item: Codable & Equatable>: ObservableObject {
+    ///
+    /// The user can ``try await store.setupTask.value`` to make sure the ``items``
+    /// have loaded when using the synchronous version of the ``init``.
+    public private(set) lazy var setupTask: Task<Void, Error> = Task { @MainActor in
```
This is the one thing in the patch that I'm stumped on, I don't really understand the expected usage of `setupTask` external to this class.

I see you wrote
> The user can ``try await store.setupTask.value`` to make sure the ``items`` have loaded when using the synchronous version of the ``init``

But I'm not sure in practice how this would be used. Is the expectation that a user would query `setupTask` to see if all items have been loaded, or use it to trigger a full reload, or something else that I'm not considering? If it's either of those two then perhaps a different abstraction (a name change, a different function, etc) may be the right answer, or I may be misunderstanding the intended purpose.

rl-pavel replied:
Yeah I'm right there with you, and honestly I'm not 100% sure how to make the API more clear. 

To step back a little, the current issue is that there's no way to know when the `items` have loaded (except for, afaik, `objectWillChange` observations, like in `Stored`). And storing/exposing this task opens up a couple of ways of handling that in different situations. 

My thought process is this:
1. If the `Store` is used directly and can be initialized/stored from an `async` context, then using the `async` `init` is the simplest option - that will create the store and return once `items` have loaded. 
2. If the `Store` is used directly, but cannot be initialized from an `async` context for some reason, and you still want to make sure `items` are retrieved after they've been loaded, you can do something like:
```
let store: Store<SomeItem> = .init(...) // non-async

func fetchItems() async -> [SomeItem] { 
  await store.setupTask.value
  return await store.items
}
```
3. And `@Stored` is an edge case of the 2nd point above, i.e. `Store` has to be created synchronously, except it handles changes internally using Combine. 

I'm open to other suggestions though in terms of naming/api. Maybe instead of exposing the task directly, it could be wrapped in `func itemsDidLoad() async` or something, so it'd be `await store.itemsDidLoad()` instead of `setupTask.value`?

mergesort replied:
Thank you for the explanation of what you were thinking, that definitely helps me understand the problem more clearly.

Since the goal is to communicate whether or not the items have loaded, I do like the idea of exposing a computed property of along the lines of your suggested `func itemsDidLoad() async`. I don't usually like to expose implementation details such as the underlying tasks, which is why I'm leaning towards your proposed solution of only telling the user whether the items have loaded, as I can see the benefit of knowing that in an async world.

Could we consider a name like `storeHasInitialized`, `storeHasLoaded`, or `itemsHaveLoaded`  for the property, returning the has loaded state as a `Bool`? (I prefer `storeHasInitialized` because it implies that this operation should only occur once, but I'm also _very_ open to suggestions.)

Thanks again for taking the time to help me get where you're coming from!

mergesort commented:
I apologize for any delays over the last week, I actually went on vacation a few days ago so the timing was poor, but failed to communicate that I won't be available for much of the next week with this pull request in progress.

rl-pavel commented:
> I apologize for any delays over the last week, I actually went on vacation a few days ago so the timing was poor, but failed to communicate that I won't be available for much of the next week with this pull request in progress.

All good, I actually just started a new job this week, so I might be slow to respond as well. Will try to get to your feedback this weekend!

mergesort commented:
No rush, but I wanted to say congrats on the new job! 🥳

rl-pavel commented:
@mergesort thank you! I have finally gotten back to tinkering with my side project and spent a bit of time on this as well. I think I've addressed most of your feedback:
* Made ~`setUpTask`~ > `loadStoreTask` private and exposed the completion with `itemsHaveLoaded() async throws`.
* Updated the comments for the non-async `init`s with some realistic examples of how to await the `items` to load.
* Reverted `StoreTests` to use the non-async init by wrapping the initialization into a non-async nested func.
* Added `AsyncStoreTests` with matching tests, except for using the async store init.
* Not asked, but I've found it helpful to have matching tests when using the `@Stored` wrapper, just to make sure I didn't break anything with it. Lives is `StoredTests` + I had to update `BoutiqueItem` to work with that. 

One thing I'm not sure about is the documentation. As I mentioned, I've updated the comments in the code, but I haven't worked with DocC before and couldn't work out how to update the stuff there. Lmk and I'd be happy to add my notes there as well 🙂

mergesort commented:
Thank you so much for the updates @rl-pavel! I promise that I will find the time to take a look at everything on Sunday, and assuming it all looks good (which it sounds like it should), I should be able to approve and merge it in.

---

<img width="227" alt="Screenshot 2022-12-17 at 00 09 55" src="https://user-images.githubusercontent.com/716513/208226347-49a2c721-ae3c-4e83-909a-0aff69e7e118.png">

If you're familiar with markdown, working with docc is actually very similar. The only real difference is that you can hint to Xcode that you want to link to a specific type by using two backticks around a type name like so (``Store``), which will directly link to the `Store` as a result, and link directly to different docc docs with syntax like `<doc:Using-Stores>`. 

This project is already setup for docc as you can see from the docs available at [build.ms/boutique/docs](https://build.ms/boutique/docs). I believe the only doc that needs updating would be the one titled `Using Stores`, summarizing what we've been discussing here, and how to use the new async initializer. I think it would be good to add a section under `Store Operations` before `Further Exploration, @Stored And More` titled `Sync or Async?`, which provides the overview and context we've discussed here, guiding users to making the right choice of initializer for their needs.

I know this PR has been running on for a while and unfortunately I haven't had much time of late, but I don't want to hold up a good idea from being available to others. If you don't have time for or aren't interested in writing up the documentation I can work on it next week and add you as a reviewer to make sure I touched on everything and haven't missed anything.

Thank you again, and please let me know if anything sounds off!

rl-pavel commented:
@mergesort no problem, and no rush, take as much time as you need 🙂 

Ah, I somehow missed the Documentation file and was trying to look in the `docs` folder. I added a `Sync or Async?` section as requested (great title btw), let me know if anything could be clarified.

mergesort commented:
Hey @rl-pavel, your pull request is great and I'm basically ready to close out this work and accept the pull request. I didn't make any code changes but I did rewrite some parts of the documentation. What you wrote was a fantastic foundation so I mostly reworded some of it to add some clarity for people who may not be as familiar with Boutique. (I also made some formatting changes, nothing major, mostly using 4 spaces over 2, moving private variables to the bottom of classes, eliminating whitespace, etc).

Your contribution was really valuable and comprehensive, I really appreciate you going the extra mile to add good documentation so others can benefit from your work. Not only am I grateful for your help and think these changes will provide a nice benefit for Boutique and it's users, but I also learned a few things in the process.

I do have one last ask before merging in the pull request. Since I rewrote some of the docs would you be able to take a quick look at what I wrote still captures the essence of your changes? I'd love to fix any mistakes I may have made, or if anything is unclear it'd be great to catch it before it goes out to users.

I just want to say again, thank you so much for all of your help. I have a few improvements I'm going to make to Boutique over the next couple of months but I think you've made a fundamentally important improvement here, and I didn't want to go without saying it means a lot to me personally, and others using Boutique.

rl-pavel commented on Sources/Boutique/Documentation.docc/Articles/Using Stores.md:
````diff
@@ -84,6 +84,36 @@ try await store
+}
+```
+
+The synchronous initializer is a sensible default, but if your app's needs dictate displaying data only once you've loaded all of the necessary items the asynchronous initializers are there to help.
````
Another comma. Also, this doesn't mention the `itemsHaveLoaded()` approach, which I don't mind, just thought I'd point out in case it wasn't intentional. 
```suggestion
The synchronous initializer is a sensible default, but if your app's needs dictate displaying data only once you've loaded all of the necessary items, the asynchronous initializers are there to help.
```

mergesort commented:
And it's been merged, thank you again @rl-pavel for all the help and great work! I'll integrate this into Plinky to make sure it works well enough, and then cut a new release. 😄

rl-pavel commented:
Awesome, thank you! <img src="https://emojis.slackmojis.com/emojis/images/1643516619/26603/parrot_hd.gif?1643516619" width="3%"/> Excited to see what you've been working on with Plinky. Let me know if you need any more beta testers, would love to try it out 🙂
